lib/crypto.js

```diff
@@ -662,6 +672,29 @@ function filterDuplicates(names) {
   }).sort();
 }
 
+// This implements Brad Hill's Double HMAC pattern from
+// https://www.nccgroup.trust/us/about-us/
+//   newsroom-and-events/blog/2011/february/double-hmac-verification/.
+// In short, it's near-impossible to write a reliable constant-time compare in a
+// high level language like JS, because of the many layers that can optimize
+// away attempts at being constant time.
+//
+// Double HMAC avoids that problem by blinding the timing channel instead. After
+// running the inputs through a second round of HMAC, we are free to
+// short-circuit comparison, because the time it takes to reach the
+// short-circuit has no relation to the similarity between the guessed digest
+// and the correct one.
+exports.timingSafeEqual = timingSafeEqual;
+function timingSafeEqual(a, b) {
+  var key = randomBytes(32);
+  var ah = new Hmac('sha256', key).update(a).digest();
+  var bh = new Hmac('sha256', key).update(b).digest();
+  // The test for === is just in case of the vanishingly small chance of
+  // a collision. It only fires if the digest comparison passes and so doesn't
+  // leak timing information.
+  return ah.equals(bh) && a.toString() === b.toString();
```

Comment:
Perhaps it might be better to only use `toString()` if *both* `a` and `b` are *not* Buffers, otherwise use `a.equals(b)`. That would help avoid unnecessary string creation.